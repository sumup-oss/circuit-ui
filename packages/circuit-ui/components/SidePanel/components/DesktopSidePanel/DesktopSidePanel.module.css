.base {
  height: 100%;
  outline: none;
  background-color: var(--cui-bg-normal);
  box-shadow: inset var(--cui-border-width-kilo) 0 0 var(--cui-border-divider);
  transform: translateX(100%);
  transition: transform var(--cui-transitions-slow);
}

.open {
  transform: translateX(0);
}

.instant {
  transition: none;
}

.closed {
  transform: translateX(100%);
  transition: transform var(--cui-transitions-slow);
}

.overlay {
  height: 100%;
}
