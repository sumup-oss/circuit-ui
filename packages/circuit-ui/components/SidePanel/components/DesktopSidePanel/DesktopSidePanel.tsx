/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { css, ClassNames, Global } from '@emotion/react';
import ReactModal, { Props as ReactModalProps } from 'react-modal';

import { StackContext } from '../../../StackContext';
import {
  SidePanelProps,
  SIDE_PANEL_WIDTH,
  PORTAL_CLASS_NAME,
  TRANSITION_DURATION_DESKTOP,
} from '../../SidePanel';

export type DesktopSidePanelProps = ReactModalProps &
  Pick<SidePanelProps, 'isInstantOpen' | 'top'>;

export const DesktopSidePanel = ({
  children,
  isInstantOpen,
  top,
  ...props
}: DesktopSidePanelProps): JSX.Element => (
  <ClassNames>
    {({ css: cssString, cx, theme }) => {
      // React Modal styles
      // https://reactcommunity.org/react-modal/styles/classes/
      const styles = {
        base: cssString`
          height: 100%;
          outline: none;
          background-color: var(--cui-bg-normal);
          box-shadow: inset ${theme.borderWidth.kilo} 0px 0px var(--cui-border-divider);
          transform: translateX(100%);
          transition: transform ${TRANSITION_DURATION_DESKTOP}ms ease-in-out;
        `,
        afterOpen: cx(
          cssString`
            transform: translateX(0);
          `,
          isInstantOpen &&
            cssString`
              transition: none;
            `,
        ),
        beforeClose: cssString`
          transform: translateX(100%) !important;
          transition: transform ${TRANSITION_DURATION_DESKTOP}ms ease-in-out !important;
        `,
      };

      const overlayStyles = cssString`
        height: 100%;
      `;

      const reactModalProps: ReactModalProps = {
        ariaHideApp: false,
        className: styles,
        overlayClassName: overlayStyles,
        shouldCloseOnOverlayClick: false,
        closeTimeoutMS: TRANSITION_DURATION_DESKTOP,
        ...props,
      };

      return (
        <StackContext.Provider value={theme.zIndex.absolute}>
          <ReactModal {...reactModalProps}>{children}</ReactModal>
          <Global
            styles={css`
              /* Enable keyboard navigation inside the modal, see https://github.com/reactjs/react-modal/issues/782 */
              .${PORTAL_CLASS_NAME} {
                position: fixed;
                top: ${top};
                right: 0;
                bottom: 0;
                width: ${SIDE_PANEL_WIDTH};
                z-index: ${theme.zIndex.absolute};
              }
            `}
          />
        </StackContext.Provider>
      );
    }}
  </ClassNames>
);
