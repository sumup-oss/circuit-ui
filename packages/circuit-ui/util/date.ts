/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Temporal } from 'temporal-polyfill';
import { formatDateTime } from '@sumup-oss/intl';

import type { Locale } from './i18n.js';

export type FirstDayOfWeek = 1 | 7;
export type DaysInWeek = number;
export type PlainDateRange =
  | []
  | [Temporal.PlainDate]
  | [Temporal.PlainDate, Temporal.PlainDate];

// ISO 8601 timestamps only support positive 4-digit years
export const MIN_YEAR = 1;
export const MAX_YEAR = 9999;

export const MIN_MONTH = 1;
export const MAX_MONTH = 12;

export const MIN_DAY = 1;
// MAX_DAY is not constant as it depends on the year and month

export function getTodaysDate() {
  return Temporal.Now.plainDateISO();
}

export function isPlainDate(date: unknown): date is Temporal.PlainDate {
  return date instanceof Temporal.PlainDate;
}

export function toPlainDate(date?: string): Temporal.PlainDate | undefined {
  if (!date) {
    return undefined;
  }
  try {
    return Temporal.PlainDate.from(date);
  } catch {
    return undefined;
  }
}

export function sortDateRange<T extends PlainDateRange>(dateRange: T): T {
  return dateRange.sort((a, b) => Temporal.PlainDate.compare(a, b)) as T;
}

export function clampDate(
  date: Temporal.PlainDate,
  minDate?: Temporal.PlainDate | null,
  maxDate?: Temporal.PlainDate | null,
): Temporal.PlainDate {
  if (minDate && Temporal.PlainDate.compare(date, minDate) < 0) {
    return minDate;
  }
  if (maxDate && Temporal.PlainDate.compare(date, maxDate) > 0) {
    return maxDate;
  }
  return date;
}

export function getFirstDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  const diff =
    (date.dayOfWeek < firstDayOfWeek ? date.daysInWeek : 0) +
    date.dayOfWeek -
    firstDayOfWeek;
  return date.subtract({ days: diff });
}

export function getLastDateOfWeek(
  date: Temporal.PlainDate,
  firstDayOfWeek: FirstDayOfWeek,
): Temporal.PlainDate {
  return getFirstDateOfWeek(date, firstDayOfWeek).add({
    days: date.daysInWeek - 1,
  });
}

export function getMonthName(month: number, locale?: Locale) {
  // The year can be arbitrary since the month names are the same every year
  const yearMonth = new Temporal.PlainYearMonth(2000, month, 'gregory');
  return formatDateTime(yearMonth, locale, { month: 'long' });
}
