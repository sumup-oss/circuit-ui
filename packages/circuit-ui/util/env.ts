/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export function getEnvVariable(name: string) {
  if (name === 'NODE_ENV') {
    // Some bundlers have special logic for `process.env.NODE_ENV` which
    // relies on it being written as a continuous string. Destructuring or
    // dynamically accessing `NODE_ENV` can break this logic.
    throw new Error('Do not dynamically access NODE_ENV');
  }

  if (typeof process !== 'undefined') {
    return process.env?.[name];
  }

  return undefined;
}
