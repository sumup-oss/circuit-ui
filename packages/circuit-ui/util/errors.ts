/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* eslint-disable max-classes-per-file */
import React from 'react';

export class CircuitError extends Error {
  constructor(componentName: string, message: string) {
    super(`[${componentName}] ${message}`);
    this.name = 'CircuitError';
    // Adapted from https://stackoverflow.com/questions/33474179/react-access-parent-component-name

    /* eslint-disable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactCurrentOwner &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner
        .return.type
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
          .ReactCurrentOwner.return.type as string;
    }
    if (
      // @ts-expect-error Since this code only runs in development, it's fine to use this internal React API.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame &&
      // @ts-expect-error This is fine.
      React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED
        .ReactDebugCurrentFrame.getCurrentStack
    ) {
      this.stack =
        // @ts-expect-error This is fine.
        React.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactDebugCurrentFrame.getCurrentStack() as string;
    }
    /* eslint-enable max-len, no-underscore-dangle, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-unsafe-call */
  }
}

export class DeprecationError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'DeprecationError';
  }
}

export class AccessibilityError extends CircuitError {
  constructor(componentName: string, message: string) {
    super(componentName, message);
    this.name = 'AccessibilityError';
  }
}
