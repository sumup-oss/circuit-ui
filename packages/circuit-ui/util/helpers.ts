/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isArray, isFunction, isObject, isString } from './type-check';

/**
 * Calls each function in an array with the arguments it receives.
 * Falsy functions are ignored.
 */
export function eachFn<Args extends []>(
  fns: (undefined | ((...args: Args) => unknown))[],
) {
  return (...args: Args): void =>
    fns.forEach((fn) => isFunction(fn) && fn(...args));
}

/**
 * Checks whether an array, object, or string is empty or whether the value is
 * falsy.
 */
export function isEmpty(value: unknown): boolean {
  if (!value) {
    return true;
  }

  if (isString(value) || isArray(value)) {
    return !value.length;
  }

  if (isObject(value)) {
    return !Object.keys(value).length;
  }

  return false;
}

export function clamp(value: number, min: number, max: number): number {
  if (process.env.NODE_ENV !== 'production' && min >= max) {
    throw new RangeError(
      `The minimum value (${min}) must be less than the maximum value (${max}).`,
    );
  }
  return Math.max(min, Math.min(value, max));
}

type Fn<T extends []> = (...args: T) => void;

/**
 * Triggers a function at most once in a given amount of time.
 */
export function throttle<T extends []>(fn: Fn<T>, timeout: number): Fn<T> {
  let ready = true;
  return (...args: T) => {
    if (!ready) {
      return;
    }

    ready = false;
    fn(...args);
    setTimeout(() => {
      ready = true;
    }, timeout);
  };
}
