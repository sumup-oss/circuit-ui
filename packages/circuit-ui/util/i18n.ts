/**
 * Copyright 2024, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { isString } from './type-check.js';

export const FALLBACK_LOCALE = 'en-US';

export const SUPPORTED_LOCALES = [
  'bg-BG',
  'cs-CZ',
  'da-DK',
  'de-AT',
  'de-CH',
  'de-DE',
  'de-LU',
  'el-CY',
  'el-GR',
  'en-AU',
  'en-GB',
  'en-IE',
  'en-MT',
  'en-US',
  'es-CL',
  'es-CO',
  'es-ES',
  'es-MX',
  'es-PE',
  'es-US',
  'et-EE',
  'fi-FI',
  'fr-BE',
  'fr-CH',
  'fr-FR',
  'fr-LU',
  'hr-HR',
  'hu-HU',
  'it-CH',
  'it-IT',
  'lt-LT',
  'lv-LV',
  'nb-NO',
  'nl-BE',
  'nl-NL',
  'pl-PL',
  'pt-BR',
  'pt-PT',
  'ro-RO',
  'sk-SK',
  'sl-SI',
  'sv-SE',
] as const;

export type Locale = string | string[];
type SupportedLocale = (typeof SUPPORTED_LOCALES)[number];

export type Translations<Key extends string | number | symbol> = Record<
  SupportedLocale,
  Record<Key, string>
>;

/**
 * Returns the user's preferred locale(s) in browser-like environments.
 */
export function getDefaultLocale(): Locale {
  if (typeof window === 'undefined') {
    return FALLBACK_LOCALE;
  }
  return (navigator.languages ||
    navigator.language ||
    FALLBACK_LOCALE) as Locale;
}

/**
 * Returns the first supported locale.
 */
export function findSupportedLocale(locale: Locale): SupportedLocale {
  const locales = isString(locale) ? [locale] : locale;

  // eslint-disable-next-line no-restricted-syntax
  for (const l of locales) {
    const matcher =
      locale.length === 5
        ? (value: string) => value === l
        : (value: string) => value.startsWith(l);

    const match = SUPPORTED_LOCALES.find(matcher);

    if (match) {
      return match;
    }
  }

  return FALLBACK_LOCALE;
}

export function transformModulesToTranslations<
  T extends Record<string, string>,
  Key extends string | number | symbol = keyof T,
>(modules: Record<string, T>): Translations<Key> {
  const translations = Object.entries(modules).reduce(
    (acc, [importPath, exports]) => {
      const { default: unused, ...strings } = exports;
      const matches = importPath.match(/[a-z]{2}-[A-Z]{2}/);

      // @ts-expect-error This environment variable is set by Vite.
      if (import.meta.env.DEV && !matches) {
        throw new Error(
          `Failed to extract a locale from the import path: ${importPath}`,
        );
      }

      // biome-ignore lint/style/noNonNullAssertion:
      const locale = matches![0] as SupportedLocale;

      // @ts-expect-error This environment variable is set by Vite.
      if (import.meta.env.DEV && !SUPPORTED_LOCALES.includes(locale)) {
        throw new Error(`Unsupported locale: ${importPath}`);
      }

      acc[locale] = strings as unknown as Record<Key, string>;
      return acc;
    },
    {} as Translations<Key>,
  );

  // @ts-expect-error This environment variable is set by Vite.
  if (import.meta.env.DEV) {
    SUPPORTED_LOCALES.forEach((locale) => {
      if (!translations[locale]) {
        throw new Error(`Missing translations for locale ${locale}`);
      }
    });
  }

  return translations;
}
