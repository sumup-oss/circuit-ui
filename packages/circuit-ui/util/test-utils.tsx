/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { FunctionComponent, ReactElement } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, RenderResult } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import {
  ComponentsContext,
  defaultComponents,
} from '../components/ComponentsContext/ComponentsContext';

export * from '@testing-library/react';

export type RenderFn<T = any> = (component: ReactElement, ...rest: any) => T;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const render: RenderFn<RenderResult> = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn<string> = (component) =>
  renderToStaticMarkup(<WithProviders>{component}</WithProviders>);
const create = (...args: Parameters<RenderFn<RenderResult>>) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export { create, render, renderToHtml, renderHook, actHook, userEvent, axe };
