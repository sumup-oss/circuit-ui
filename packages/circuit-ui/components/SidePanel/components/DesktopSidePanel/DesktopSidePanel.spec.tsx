/**
 * Copyright 2022, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render, axe, act } from '../../../../util/test-utils';

import { DesktopSidePanel, DesktopSidePanelProps } from './DesktopSidePanel';

describe('DesktopSidePanel', () => {
  const baseProps: DesktopSidePanelProps = {
    isInstantOpen: false,
    top: '0px',
    // Silences the warning about the missing app element.
    // In user land, the side panel is always rendered by the SidePanelProvider,
    // which takes care of setting the app element.
    // http://reactcommunity.org/react-modal/accessibility/#app-element
    ariaHideApp: false,
    // Keep the modal opened by setting the react-modal isOpen prop.
    // Usually this is controlled by the SidePanelProvider.
    isOpen: true,
  };

  const renderComponent = (props: Partial<DesktopSidePanelProps> = {}) =>
    render(<DesktopSidePanel {...baseProps} {...props} />);

  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should match the snapshot', () => {
    const { baseElement } = renderComponent();
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the side panel', () => {
    const { getByRole } = renderComponent();
    expect(getByRole('dialog')).toBeVisible();
  });

  it('should slide in from the right', () => {
    const { getByRole } = renderComponent();
    expect(getByRole('dialog')).toHaveStyle('transform: translateX(100%)');
  });

  it('should open without animation when isInstantOpen is true', () => {
    const { getByRole } = renderComponent({ isInstantOpen: true });
    act(() => {
      jest.advanceTimersByTime(100);
    });
    expect(getByRole('dialog')).toHaveStyle('transition: none');
  });

  it('should describe the side panel as modal', () => {
    const { getByRole } = renderComponent();
    expect(getByRole('dialog')).toHaveAttribute('aria-modal', 'true');
  });

  /**
   * FIXME: calling axe here triggers an act() warning.
   */
  it('should meet accessibility guidelines', async () => {
    jest.useRealTimers();
    const { container } = renderComponent();
    const actual = await axe(container);
    expect(actual).toHaveNoViolations();
  });
});
