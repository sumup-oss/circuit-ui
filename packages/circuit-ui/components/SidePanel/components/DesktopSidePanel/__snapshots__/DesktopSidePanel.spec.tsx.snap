// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DesktopSidePanel should match the snapshot 1`] = `
.circuit-0 {
  height: 100%;
}

.circuit-1 {
  height: 100%;
  outline: none;
  background-color: var(--cui-bg-normal);
  box-shadow: inset 1px 0px 0px var(--cui-border-divider);
  -webkit-transform: translateX(100%);
  -moz-transform: translateX(100%);
  -ms-transform: translateX(100%);
  transform: translateX(100%);
  -webkit-transition: -webkit-transform 200ms ease-in-out;
  transition: transform 200ms ease-in-out;
}

<body
  class="ReactModal__Body--open"
>
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div />
  <div
    data-react-modal-body-trap=""
    style="position: absolute; opacity: 0;"
    tabindex="0"
  />
  <div
    class="ReactModalPortal"
  >
    <div
      class="ReactModal__Overlay circuit-0"
    >
      <div
        aria-modal="true"
        class="circuit-1"
        role="dialog"
        tabindex="-1"
      />
    </div>
  </div>
</body>
`;
